packages/shared/src/api/chatApi.ts

```diff
@@ -1,8 +1,76 @@
+import { Poke, unixToDa } from '@urbit/api';
+
 import * as db from '../db';
+import { JSONToInlines } from '../logic/tiptap';
 import * as ub from '../urbit';
-import { toClientMeta } from './converters';
+import {
+  ClubAction,
+  ClubDelta,
+  DmAction,
+  JSONContent,
+  WritDelta,
+  WritDeltaAdd,
+  WritDiff,
+  WritResponse,
+  constructStory,
+  whomIsDm,
+} from '../urbit';
+import { formatUd, toClientMeta } from './converters';
 import { poke, scry } from './urbit';
 
+interface OptimisticAction {
+  action: Poke<DmAction | ClubAction>;
+  event: WritDiff | WritResponse;
+}
+
+function dmAction(ship: string, delta: WritDelta, id: string): Poke<DmAction> {
+  return {
+    app: 'chat',
+    mark: 'chat-dm-action',
+    json: {
+      ship,
+      diff: {
+        id,
+        delta,
+      },
+    },
+  };
+}
+
+function multiDmAction(id: string, delta: ClubDelta): Poke<ClubAction> {
+  return {
+    app: 'chat',
+    mark: 'chat-club-action-0',
+    json: {
+      id,
+      diff: {
+        uid: '0v3',
+        delta,
+      },
+    },
+  };
+}
+
+function getActionAndEvent(
```

Comment:
maybe we collapse these into a single action builder since it seems like we never use the event/diff?